### PR TITLE
Theming: Effects (Depth and rounded corners)

### PR DESCRIPTION
**Pull request by phkuo**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [x] Include a change request file using `$ npm run change`

#### Description of changes

Add an experimental API to theming for upcoming Fluent work, which adds shadows and rounded corners.

### CODE REVIEW

File: packages/styling/src/interfaces/ISerifs.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { IRawStyle } from '@uifabric/merge-styles';
+
+/**
+ * @internal
+ * Experimental interface for decorative styling in a theme.
+ */
+export interface ISerifs {
```
Comment (phkuo) [inline packages/styling/src/interfaces/ISerifs.ts]: They say the single hardest thing in computer science is naming things and off by one errors.

I'm not thrilled with the name here, but it gets the point across. Extra, decorative styles that don't strongly influence functionality but help things feel better. "Flair" doesn't sound quite right, and "styles" or "visualizations" seem to vague.

File: packages/styling/src/interfaces/ISerifs.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  depth4: IRawStyle;
+  depth8: IRawStyle;
+  depth16: IRawStyle;
```
Comment (phkuo) [inline packages/styling/src/interfaces/ISerifs.ts]: There doesn't seem to be a depth32 step, not sure if that's intentional or part of Fluent design.

File: packages/styling/src/styles/DefaultSerifs.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { ISerifs } from '../interfaces/index';
+
+export const DefaultSerifs: ISerifs = {
+  depth4: { boxShadow: '0 0 0 0 transparent' },
+  depth8: { boxShadow: '0 3.2px 7.2px 0 rgba(0, 0, 0, 0.132), 0 0.6px 1.8px 0 rgba(0, 0, 0, 0.108)' },
+  depth16: { boxShadow: '0 3.2px 7.2px 0 rgba(0, 0, 0, 0.132), 0 0.6px 1.8px 0 rgba(0, 0, 0, 0.108)' },
+  depth64: { boxShadow: '0 25.6px 57.6px 0 rgba(0, 0, 0, 0.22), 0 4.8px 14.4px 0 rgba(0, 0, 0, 0.18)' },
```
Comment (phkuo) [inline packages/styling/src/styles/DefaultSerifs.ts]: Depth values are copied from: packages\fluent-theme\src\fluent\FluentDepths.ts

File: packages/styling/src/styles/DefaultSerifs.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+  roundedCorner2: 2,
+  roundedCorner4: 4
```
Comment (phkuo) [inline packages/styling/src/styles/DefaultSerifs.ts]: Rounded corner values are made-up for now, I couldn't find the real values.

File: packages/styling/src/interfaces/ISerifs.ts (hunk 1 above)
Comment (joschect) [reply packages/styling/src/interfaces/ISerifs.ts]: Doesn't Serif almost always have to do with font? This seems a bit misleading.

File: packages/styling/src/interfaces/ISerifs.ts (hunk 2 above)
Comment (mikewheaton) [reply packages/styling/src/interfaces/ISerifs.ts]: That's intentional. :)

File: packages/styling/src/styles/DefaultSerifs.ts (hunk 4 above)
Comment (mikewheaton) [reply packages/styling/src/styles/DefaultSerifs.ts]: I've only seen 2px rounded corners in the Fluent designs so far.

File: packages/styling/src/styles/DefaultSerifs.ts (hunk 3 above)
Comment (mikewheaton) [reply packages/styling/src/styles/DefaultSerifs.ts]: Is it possible to import these from the Fluent theme instead of copying them? While the values are unlikely to change, these could get out of sync in the future.

File: packages/styling/src/interfaces/ISerifs.ts (hunk 1 above)
Comment (mikewheaton) [reply packages/styling/src/interfaces/ISerifs.ts]: I'm also confused by what a 'serif' is in this context. I would have assumed it was something font-related.

Comment (Jahnp) [reply packages/styling/src/interfaces/ISerifs.ts]: Agreed. Naming things is tough, but serif isn't right here, especially since themes also support fonts. 

What about `IDecorations`?

File: packages/styling/src/styles/DefaultSerifs.ts (hunk 3 above)
Comment (Jahnp) [reply packages/styling/src/styles/DefaultSerifs.ts]: +1

It'd be great to not duplicate values here.

File: packages/styling/src/interfaces/ISerifs.ts (hunk 1 above)
Comment (dzearing) [reply packages/styling/src/interfaces/ISerifs.ts]: Are these `elevations`?

File: packages/styling/src/interfaces/ISerifs.ts (hunk 2 above)
Comment (dzearing) [reply packages/styling/src/interfaces/ISerifs.ts]: I'd really rather not use 4, 8, 16, etc. Either:

we snap to 100 scale; (depth100, depth200, etc)

or, we do small/medium/large/etc

File: packages/styling/src/interfaces/ISerifs.ts (hunk 1 above)
Comment (Jahnp) [reply packages/styling/src/interfaces/ISerifs.ts]: @dzearing -- I think these are meant to be other kinds of customizations that aren't fonts or colors.

File: packages/styling/src/styles/DefaultSerifs.ts (hunk 3 above)
Comment (phkuo) [reply packages/styling/src/styles/DefaultSerifs.ts]: I intentionally copied them instead of importing, since the theme should be the "source of truth". Eventually, once we finish the Fluentization work, that project will be going away anyways, right?
(Also, shadows can be added in parallel to the Fluentization work. Unlike rounded corners and colors, it can coexist with old stuff.)

**Comment by phkuo** [comment]
@Jahnp @dzearing @joschect @mikewheaton 
New proposals for naming, what do you think of these?

ISerifs -> IDecorations, IAdornments, or IAffordances (though technically not correct, I don't think rounded corners count as an affordance)

depth4/8/16/64 -> shadow100, 200, 400, 1600 (so that the numbers are roughly proportional to the amount of shadow). Naming them `shadow` gives more indication of what they do. `layer` or `level` may work also.
@dzearing I feel like the 100/200/400... for shadows is excessive. The values are very small already, it's possible we add a 150 step in between, but highly unlikely we'll need a 110 and a 105 and a 101 step given that those would not even be perceptible differences.

Remove roundedCorner4... will leave in roundedCorner2 with the number in case we want to add more steps in the future.

File: packages/styling/src/interfaces/ISerifs.ts (hunk 1 above)
Comment (betrue-final-final) [reply packages/styling/src/interfaces/ISerifs.ts]: Yes the single most confusing PR name I've ever seen. @phkuo you know you can ask a designer for help! May I suggest Surface styling?

File: packages/styling/src/interfaces/ISerifs.ts (hunk 2 above)
Comment (betrue-final-final) [reply packages/styling/src/interfaces/ISerifs.ts]: Note that 4, 8, 16 etc are the names used in Fluent across the company today.

Comment (betrue-final-final) [reply packages/styling/src/interfaces/ISerifs.ts]: ISerifs?????? Ok that is a damaging name. Someone in the future is going to waste a lot of time trying to figure that out in the code. Let's pick a different name please?

File: packages/styling/src/styles/DefaultSerifs.ts (hunk 4 above)
Comment (betrue-final-final) [reply packages/styling/src/styles/DefaultSerifs.ts]: 2px is correct.

File: packages/styling/src/interfaces/ISerifs.ts (hunk 2 above)
Comment (betrue-final-final) [reply packages/styling/src/interfaces/ISerifs.ts]: How about ISurfaces, or ISurfaceStyles?

File: packages/styling/src/interfaces/ISerifs.ts (hunk 1 above)
Comment (dzearing) [reply packages/styling/src/interfaces/ISerifs.ts]: @Jahnp I'm unclear then on the use case. Can you or @phkuo provide an example of the use case?

Comment (phkuo) [reply packages/styling/src/interfaces/ISerifs.ts]: @dzearing For theming shadows and rounded corners. For example, if you want to use a glow instead of a box-shadow, or if you want sharp corners instead of rounded corners. For now, we would set the default roundedCorner value to 0, and in the test Fluent theme, set it to 2.

**Comment by betrue-final-final** [comment]
Ok @phkuo and I rummaged around art tools and such for a name to call this, and we landed on IEffects. Shadows, rounded corners, opacity are all "Effects", especially if you consider that we have precedence in Office.
image

**Comment by dzearing** [comment]
I'm still unsure of the actual use cases here. For example, let's just focus only on shadowing.

Are we planning to:

1. Add stops for just shadowing?
2. Update all places where shadowing is used to use it?
3. Moving on to other specific treatments such as shadowing and glow and repeat #1 and 2?

Or, are we just adding a grab bag of "effects" that will be scattered?

I'd prefer to really understand what the steps are here before we commit to it.

**Comment by dzearing** [comment]
@JasonGore heads up.

Given these are listed as internal I think I'm ok with this. Philip and I talked offline, and I do think it could be nice to have a generalized way to define standard effects which have some sort of meaning, from the theme. But it certainly feels like palette colors and would need some sort of abstraction layer to define which things use which effects.

I wonder if the same thing could be done with customizations as we have them today.

For example you could define standard IRawStyle definitions in a library:

```ts
const Elevation2: IRawStyle = { ... };
const Elevation4: IRawStyle = { ... };
etc;
``

... and then using those via Customizer targeting things:

```ts
Button: {
  styles: {
     root: Elevation2
  }
}
```

... but then, there are friction points here - you need to know that root is the thing that should take the dropshadow (which might not be the case), plus you might have specificity wars to deal with. A better approach would be to expose knobs for common configurable things:

```ts
Button: {
  styleVariables : {
    shadow: ShadowElevation2
  }
}
```

And then let the styling inject the shadow in the right place. I think.

File: packages/styling/src/interfaces/IEffects.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+ * @internal
+ * Experimental interface for decorative styling in a theme.
+ */
+export interface IEffects {
+  elevation4: IRawStyle;
+  elevation8: IRawStyle;
+  elevation16: IRawStyle;
+  elevation64: IRawStyle;
+
+  roundedCorner2: number;
```
Comment (phkuo) [inline packages/styling/src/interfaces/IEffects.ts]: I made this a number so you can do math on it. (For example, there's a really big button so you want double the border-radius.) Is that a useful use case or should I just make it a string? Having it as a number also limits it to pixel values only... which generally makes sense. You probably never want to have a % border-radius value globally.

File: packages/styling/src/interfaces/IEffects.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  /**
+   * Used to provide a visual affordance that this element is elevated above the surface it rests on.
+   * Usually a shadow. This is lower than elevations with a higher value, and higher than elevations with a lower value.
+   * Used for: search result dropdowns, hover cards, tooltips, help bubbles
+   */
+  elevation16: IRawStyle;
+  /**
+   * Used to provide a visual affordance that this element is elevated above the surface it rests on.
+   * Usually a shadow. This is lower than elevations with a higher value, and higher than elevations with a lower value.
+   * Used for: panels, dialogs
```
Comment (Jahnp) [inline packages/styling/src/interfaces/IEffects.ts]: Very, very small nit: Would be nice to properly capitalize component names (e.g. panels -> Panels)

File: packages/styling/src/interfaces/IEffects.ts (hunk 5 above)
Comment (JasonGore) [reply packages/styling/src/interfaces/IEffects.ts]: How will this interface grow over time? Per David's example, I'm a bit concerned that if `IEffects` applies to a `shadow` style variable, for example, that typings would allow `shadow: roundedCorner2` without necessarily making sense. Does this type need to be broken down more?

File: packages/styling/src/styles/DefaultEffects.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import { IEffects } from '../interfaces/index';
+
+export const DefaultEffects: IEffects = {
+  elevation4: { boxShadow: '0 0 5px 0 rgba(0,0,0,.4)' }, // '0 1.6px 3.6px 0 rgba(0, 0, 0, 0.132), 0 0.3px 0.9px 0 rgba(0, 0, 0, 0.108)'
+  elevation8: { boxShadow: '0 0 5px 0 rgba(0,0,0,.4)' }, // '0 3.2px 7.2px 0 rgba(0, 0, 0, 0.132), 0 0.6px 1.8px 0 rgba(0, 0, 0, 0.108)'
+  elevation16: { boxShadow: '0 0 5px 0 rgba(0,0,0,.4)' }, // '0 6.4px 14.4px 0 rgba(0, 0, 0, 0.132), 0 1.2px 3.6px 0 rgba(0, 0, 0, 0.108)'
+  elevation64: { boxShadow: '0 0 5px 0 rgba(0,0,0,.4)' }, // '0 25.6px 57.6px 0 rgba(0, 0, 0, 0.22), 0 4.8px 14.4px 0 rgba(0, 0, 0, 0.18)'
```
Comment (JasonGore) [inline packages/styling/src/styles/DefaultEffects.ts]: Per our chat, I think a brief descriptive comment that commented values are Fluent values that will eventually be default may reduce confusion for anybody who comes across this later (as I was wondering why the differed.)

File: packages/styling/src/interfaces/IEffects.ts (hunk 5 above)
Comment (phkuo) [reply packages/styling/src/interfaces/IEffects.ts]: That wouldn't be possible since a number isn't a valid type for box-shadow. You could, however, use roundedCorner2 for border-width or other random things. I think that's fine, since we might use font size for a padding value somewhere, to make room for text that will be rendered later, or to align text with another element.

Anyways, that's part of my motivation for making the elevation styles IRawStyles, since they would be so specific to box-shadow, that we might as well do that for the user. (Additionally, this lets you do fancy things like insert an image to use as the shadow, or mess with borders/outlines, or use a transform or glow effect, and so on.)

**Comment by KevinTCoughlin** [comment]
@phkuo I think you just need to run `npm run update-api` for CI to pass here.

```shell
FAILURE (1)
================================
office-ui-fabric-react (1 minute 10.1 seconds)
You have changed the public API signature for this project. Please overwrite etc/office-ui-fabric-react.api.ts with a copy of temp/office-ui-fabric-react.api.ts and then request an API review. See the Git repository README.md for more info.
================================
```
